crates/wasi-common/src/sys/unix/hostcalls_impl/misc.rs

```diff
@@ -126,9 +126,21 @@ fn poll_oneoff_handle_fd_event<'a>(
     ready_events: impl Iterator<Item = (FdEventData<'a>, yanix::poll::PollFd)>,
     events: &mut Vec<wasi::__wasi_event_t>,
 ) -> Result<()> {
+    use crate::fdentry::Descriptor;
     use std::{convert::TryInto, os::unix::prelude::AsRawFd};
     use yanix::{file::fionread, poll::PollFlags};
 
+    fn query_nbytes(fd: &Descriptor) -> Result<usize> {
+        // fionread may overflow for large files, so use another way for regular files.
+        if let Descriptor::OsHandle(os_handle) = fd {
+            let meta = os_handle.metadata()?;
+            if meta.file_type().is_file() {
+                return meta.len().try_into().map_err(Into::into);
```

Comment:
`query_nbytes`, and `fionread` too while we're here, should return `u64`, so that they work consistently on 32-bit hosts with large files.

This code returns the size of the file, however if the application has already started reading, the current file position should be subtracted from the file size.

Using the file size assumes that all the file data is immediately available, which is unfortunate. However, that is what Unix traditionally does, so it seems reasonable to do it here too, but it would be good to add a comment noting this assumption.

Reply:
> `query_nbytes`, and `fionread` too while we're here, should return `u64`, so that they work consistently on 32-bit hosts with large files.

That's true for `query_nbytes` but `fionread` should probably return `u32`, because it's return value is inherently 32-bit. 

> This code returns the size of the file, however if the application has already started reading, the current file position should be subtracted from the file size.

Good catch! Thanks.

I address your feedback.